tutorials/roofit/rf616_morphing.py

```diff
@@ -0,0 +1,138 @@
+## \file
+## \ingroup tutorial_roofit
+## \notebook
+## Use Morphing in RooFit.
+##
+## This tutorial shows how to use template morphing inside RooFit. As input we have several
+## Gaussian distributions. The output is one gaussian, with a specific mean value.
+## Since likelihoods are often used within the framework of morphing, we provide a
+## way to estimate the negative log likelihood (nll).
+##
+## Based on example of Kyle Cranmer https:##gist.github.com/cranmer/46fff8d22015e5a26619.
+## \macro_code
+## \macro_output
+##
+## \date August 2024
+## \author Robin Syring
+
+
+import ROOT
+import numpy as np
+
+# Number of samples to fill the histograms
+n_samples = 1000
+
+
+# morphing as a baseline
+def morphing(setting):
+    # set up a frame for plotting
+    frame1 = x_var.frame()
+
+    # define binning for morphing
+    bin_mu_x = ROOT.RooBinning(4, 0.0, 4.0)
+    grid = ROOT.RooMomentMorphFuncND.Grid2(bin_mu_x)
+    x_var.setBins(50)
+
+    # number of 'sampled' gaussians, if you change it, adjust the binning properly
+    parampoints = np.arange(5)
+
+    for i in parampoints:
+        # define the sampled gausians
+        mu_help = ROOT.RooRealVar(f"mu{i}", f"mu{i}", i)
+        help = ROOT.RooGaussian(f"g{i}", f"g{i}", x_var, mu_help, sigma_var)
+        workspace.Import(help)
+
+        # Fill the histograms
+        hist = workspace[f"g{i}"].generateBinned([x_var], n_samples)
+
+        # Make sure that every bin is filled and we don't get zero probability
+        for i_bin in range(hist.numEntries()):
+            hist.add(hist.get(i_bin), 1.0)
+
+        # Add the pdf to the workspace, the inOrder of 1 is necessary for calculation of the nll
+        # Adjust it to 0 to see binning
+        workspace.Import(ROOT.RooHistPdf(f"histpdf{i}", f"histpdf{i}", [x_var], hist, intOrder=0))
+
+        # Add the pdf to the grid and to the plot
+        grid.addPdf(workspace[f"histpdf{i}"], int(i))
+        workspace[f"histpdf{i}"].plotOn(frame1)
+
+    # Create the morphing and add it to the workspace
+    morph_func = ROOT.RooMomentMorphFuncND("morph_func", "morph_func", [mu_var], [x_var], grid, setting)
+
+    # Normalizes the morphed object to be a pdf, set it false to prevent warning messages and gain computational speed up
+    # morph_func.setPdfMode()
+
+    # Creating the morphed pdf
+    morph = ROOT.RooWrapperPdf("morph", "morph", morph_func, True)
+    workspace.Import(morph)
+    workspace["morph"].plotOn(frame1, LineColor="r")
+
+    c0 = ROOT.TCanvas()
+    frame1.Draw()
+
+    # The input prevents the plot from beeing closed instantly
+    input("")
```

Comment:
Tutorials should now require user input by default, because we also run them as tests etc